Dockerfile

```diff
@@ -1,31 +1,47 @@
-FROM alpine:3.9
+################################################################################
+# Builder                                                                      #
+################################################################################
+FROM alpine:3.9 as builder
 
 MAINTAINER Alexey Kutepov <reximkut@gmail.com>
 
-VOLUME ["/tmp/hypernerd"]
-
-RUN apk add --no-cache openssl-dev
-RUN apk add --no-cache curl
-RUN apk add --no-cache ghc
-RUN apk add --no-cache cabal
-RUN apk add --no-cache gmp-dev
-RUN apk add --no-cache build-base
-RUN apk add --no-cache zlib-dev
-
-RUN cabal update
-RUN mkdir /hypernerd/
-COPY ./HyperNerd.cabal /hypernerd/
-COPY ./Setup.hs /hypernerd/
-WORKDIR /hypernerd/
-RUN cabal sandbox init
+# Have base OS package installation in a single layer
+RUN \
+  apk update && \
+  apk add --no-cache \
+    openssl-dev curl ghc cabal \
+    gmp-dev build-base zlib-dev && \
+  cabal update && \
+  mkdir -p /build/
+
+WORKDIR /build/
+COPY . /build/
+
+# Install necessary dependencies and build HyperNerd
 # NOTE: happy package provides an executable that is needed to build
 # qm-interpolated-string package. qm-interpolated-string does not
 # depend on happy for some reason and that's why we have to install it
 # separately.
-RUN cabal install happy-1.19.9
-RUN cabal install --only-dependencies
-COPY ./resources/ /hypernerd/resources/
-COPY ./src/ /hypernerd/src/
-RUN cabal build exe:HyperNerd
+RUN \
+  cabal install happy-1.19.9 && \
+  cabal install -j4 --only-dependencies && \
+  cabal build exe:HyperNerd
+
+
+################################################################################
+# Application                                                                  #
+################################################################################
+FROM alpine:3.9 AS app
+
+RUN \
+  apk update && \
+  apk add --no-cache \
+    gmp libffi zlib openssl curl && \
+  mkdir -p /app
+
+COPY --from=builder /build/dist/build/HyperNerd/HyperNerd /app
+WORKDIR /app
 
-CMD ["/hypernerd/dist/build/HyperNerd/HyperNerd", "/tmp/hypernerd/secret.ini", "/tmp/hypernerd/database.db", "/tmp/hypernerd/markov.csv"]
+# TODO: Use `ENTRYPOINT` instead and instead pass arguments;  This won't work with the current setup.  Also
```

Comment:
Please format TODO according to [Snitch Format](https://github.com/tsoding/snitch#todo-body). First line is the Issue Title, second line is the Issue Description.